packages/react/src/I18nProvider.tsx

```diff
@@ -32,63 +32,58 @@ export const I18nProvider: FunctionComponent<I18nProviderProps> = ({
   forceRenderOnLocaleChange = true,
   children,
 }) => {
+  const latestKnownLocale = React.useRef<string | undefined>(i18n.locale)
   /**
    * We can't pass `i18n` object directly through context, because even when locale
    * or messages are changed, i18n object is still the same. Context provider compares
-   * reference identity and suggested workaround is create a wrapper object every time
+   * reference identity and suggested workaround is to create a wrapper object every time
    * we need to trigger re-render. See https://reactjs.org/docs/context.html#caveats.
    *
    * Due to this effect we also pass `defaultComponent` in the same context, instead
    * of creating a separate Provider/Consumer pair.
    *
    * We can't use useMemo hook either, because we want to recalculate value manually.
    */
-  const makeContext = () => ({
-    i18n,
-    defaultComponent,
-  })
-  const getRenderKey = () => {
-    return (
-      forceRenderOnLocaleChange ? i18n.locale || "default" : "default"
-    ) as string
-  }
+  const makeContext = React.useCallback(
+    () => ({
+      i18n,
+      defaultComponent,
+    }),
+    [i18n, defaultComponent]
+  )
 
-  const [context, setContext] = React.useState<I18nContext>(makeContext()),
-    [renderKey, setRenderKey] = React.useState<string>(getRenderKey())
+  const [context, setContext] = React.useState<I18nContext>(makeContext())
 
   /**
    * Subscribe for locale/message changes
    *
    * I18n object from `@lingui/core` is the single source of truth for all i18n related
    * data (active locale, catalogs). When new messages are loaded or locale is changed
    * we need to trigger re-rendering of LinguiContext.Consumers.
-   *
-   * We call `setContext(makeContext())` after adding the observer in case the `change`
-   * event would already have fired between the inital renderKey calculation and the
-   * `useEffect` hook being called. This can happen if locales are loaded/activated
-   * async.
    */
   React.useEffect(() => {
-    const unsubscribe = i18n.on("change", () => {
+    if (!forceRenderOnLocaleChange) {
+      return
+    }
+    const updateContext = () => {
+      latestKnownLocale.current = i18n.locale
       setContext(makeContext())
-      setRenderKey(getRenderKey())
-    })
-    if (renderKey === "default") {
-      setRenderKey(getRenderKey())
     }
-    if (forceRenderOnLocaleChange && renderKey === "default") {
-      console.log(
-        "I18nProvider did not render. A call to i18n.activate still needs to happen or forceRenderOnLocaleChange must be set to false."
-      )
```

Comment:
do you think users should not be warned about that case? I think it's quite useful and quickly points to the problem.

Reply:
added back